File: 153-find-minimum-in-rotated-sorted-array/153-find-minimum-in-rotated-sorted-array.py
import math
class Solution(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        # edge cases
        if len(nums) == 1 or nums[0] < nums[len(nums)-1]:
            return nums[0]

        # rotated case
        # use binary search
        
        l = 0
        r = len(nums)-1
        
        while r >= l:
            mid = (r-l)//2 + l
            if nums[mid] > nums[mid+1]:
                return nums[mid+1]
            
            if nums[mid] < nums[mid-1]:
                return nums[mid]
            
            if nums[0] > nums[mid]:
                r = r - 1
            elif nums[0] < nums[mid]:
                l = l + 1
            
        return -1
    

File: 153-find-minimum-in-rotated-sorted-array/153-find-minimum-in-rotated-sorted-array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findMin = function(nums) {
    if (nums[0] < nums[nums.length -1] || nums.length === 1) {
        return nums[0]
    }
    
    let l = 0
    let r = nums.length - 1
    while ( l <= r) {
        const mid = Math.floor((l+r)/2)
        
        if (nums[mid] > nums[mid+1]) {
            return nums[mid+1]
        } 
        
        if (nums[mid] < nums[mid - 1]) {
            return nums[mid]
        }
        
        // find out if mid is in left or right portion
        if (nums[mid] >= nums[l]) {
            l = mid + 1
        } else if (nums[mid] <= nums[r]) {
            r = mid - 1
        } 
    }
    
};